src/device/serialDevice.js

```diff
@@ -208,12 +211,31 @@ class SerialDevice extends Device {
         return Promise.resolve(cmd.length);
     }
 
+    dataLossReport() {
+        if (this.dataLossReported) return;
+        this.dataLossReported = true;
+        logger.error(
+            'Data loss detected. See https://github.com/Nordicsemiconductor/pc-nrfconnect-ppk/blob/master/doc/data-loss.md'
+        );
+    }
+
     handleRawDataSet(adcValue) {
         try {
             const currentMeasurementRange = Math.min(
                 getMaskedValue(adcValue, MEAS_RANGE),
                 this.modifiers.r.length
             );
+            const counter = getMaskedValue(adcValue, MEAS_COUNTER);
+            const expectedCounter = (this.payloadCounter + 1) & 0x3f;
+            if (this.payloadCounter !== null && expectedCounter !== counter) {
+                const diff = counter - expectedCounter;
+                const missingSamples = diff + 0x40 * (diff > 0);
```

Comment:
Is this correct? If `counter` is 12 but `expectedCounter` is 10, then `diff` is 2, which is also the number of samples we missed. But AFAICS `missingSamples` is then 0x42.

Also: I dislike algebraic terms that mix numbers and booleans. Probably because I am so bad at remembering what true and false evaluate to. So I would prefer something like (if I understand it correctly):
```js
diff >= 0 ? diff : diff + 0x40
```

Also: Please introduce a constant like `MAX_PAYLOAD_COUNTER` and use that for `0x3f` above and `0x40` here.

Reply:
Using a condition instead of algebraic booleans will be slower. With the rest I agree.

Reply:
Not sure about the performance, I never benchmarked this before. Here it seems the conditional is faster: https://jsbench.me/qgkk2eorsf

Notes:
* This of course depends on the JS engine. And this also means what is faster today may be slower tomorrow. Now the remainder was slowest in my Safari while algebraic booleans was slowest in my Brave.
* I did a different benchmark before, with a different conclusion, which did not use the result of the computation. But I was afraid that not using the result of the computation may be optimised by some JS engines to not do the computation at all, so I changed that to a loop which uses the result in the next iteration and then decreases to value to alternate between positive and negative values.

Reply:
Well, this is interesting. If you remove the named constant and just use the hardcoded value, that beats everything by quite some margin.

Reply:
You mean, remove what I called `BOUND` in JSBench? I do not see that big margins and they sadly may rather be caused by JSBench. In https://jsbench.me/enkk2sdgx5 I benched four cases:

1. `BOUND` defined in the setup
2. `BOUND` removed and used the hard coded value
3. `BOUND` defined directly above the benched code
4. `BOUND` and also the `diff` defined directly above the benched code

Surprisingly on my computer in Brave I see 2. + 3. to have about the same speed while 1 . is about 17 % slower and 4. about 21 % slower. 

On Safari (which is less relevant for Electron, just shows how different JS engines can optimise) it is different again with 1. being fastest, 4. is 9 % slower and 2. + 3. are about 19 % slower.

It is pretty hard in JavaScript to make predictions what local optimisations like these result in, especially in a different context.